### ISSUE TITLE
Enable towncrier

### ISSUE THREAD

**Issue by diazona** [diazona/setuptools-pyproject-migration#88]
We have a configuration file for [towncrier](https://towncrier.readthedocs.io/en/stable/) in the repository (inherited from https://github.com/jaraco/skeleton), so I figure we might as well make use of it. All the cool projects have a changelog, and using towncrier to maintain one is going to be slightly less annoying than doing it manually.

Actually I started working on this a little already a couple days ago while I was waiting for some other stuff, just to see how tricky it would be... seems like it's fairly straightforward. The only significant change is that every PR now has to include a changelog fragment.

@sjlongland any thoughts on this? I think if we're going to do it, best to get it out of the way now. (This should be my last last-minute addition to the initial release milestone :stuck_out_tongue:)

**Comment by diazona** [comment]
um... okay so I _may_ have spoken too soon. It's proving a little tricky to set things up to actually build the changelog when we make a release, because the existing command for building the changelog (from the skeleton, i.e. running `jaraco.develop.finalize`) seems to make some assumptions that clash with our development workflow. I'm sure it won't be too hard to figure out, I'm just not sure it's worth waiting on the initial release until we can do this.

**Comment by sjlongland** [comment]
It'd definitely be worth it if we can automate the change-log generation… at work we do it by hand, and it's the pits.  Too often, someone adds a new version but then forgets to put a link down the bottom linking to the commit changes, or mangle the URI accidentally.

https://github.com/widesky/jswidesky-client/blob/master/CHANGELOG.md is managed by hand.

That said, you don't want to try and pick up a new tool whilst flustered getting a release out.  Definitely worth a look to see if we can get it working.

**Comment by diazona** [comment]
OK sounds reasonable. I'll take a shot, I might be able to figure it out tonight... I don't think it's conceptually difficult to get this working, there are just a bunch of little pieces that need to be adjusted. (unless you want to take over? I can push my branch) And it will be nice to have this out of the way.

**Comment by sjlongland** [comment]
> OK sounds reasonable. I'll take a shot, I might be able to figure it out tonight... I don't think it's conceptually difficult to get this working, there are just a bunch of little pieces that need to be adjusted. (unless you want to take over? I can push my branch) And it will be nice to have this out of the way.

I can take over if you like… I haven't got work interfering with free time so should be able to tackle this without much interruption.

**Comment by diazona** [comment]
OK, sounds good. Though, I realized I won't be able to push changes until I get home from work anyway, so I'll spend a little bit of time on it and then push my changes, and if things haven't resolved themselves you can take over.

**Comment by diazona** [comment]
I got _something_ figured out, at least, and pushed my candidate changes as a PR (linked just above).

The code itself was not hard at all. What makes this tricky is how it affects the design of our release workflow. Back in #40 I set up the infrastructure for the following release workflow:

> 1. We start by making a Github release ("Github release" refers to the announcement page on Github that we make using [this form](https://github.com/diazona/setuptools-pyproject-migration/releases/new)), with a corresponding tag, in the Github web interface (or by using the Github API)
> 2. The workflow automatically runs and tests the code from a checkout of the main Git branch
> 3. It also tests the documentation from a checkout of the main Git branch
> 4. It builds the package for distribution
> 5. After getting a maintainer's approval through the Github UI (or API), it publishes the package to Test PyPI
> 6. Then after getting approval again, it publishes to PyPI

That procedure is designed to make the Git tag the source of truth about what version of the code actually corresponds to the release. The thing is, when we run `towncrier build` to aggregate all the changelog fragments into the actual changelog update, it needs to have the version number of the release that it's going to be building a changelog for... but the changelog needs to go into the tagged commit, which means the version number isn't available from the tag until after the changelog has been generated.

I guess the proper way to deal with this has to be that when preparing a release, we manually give towncrier the version number of the tag that's _going_ to be created. It feels a little inelegant, because towncrier is definitely capable of extracting the version from the package itself; but I think the only alternative would be changing the source of truth about the version number to be something other than tag (maybe a hard-coded constant in `pyproject.toml` or `__init__.py`? maybe the changelog itself?) and personally I really do not like doing that because it introduces ambiguity about what version of the code corresponds to what released version.

For now, just to keep things simple, I pushed a candidate change that requires us to manually run `towncrier` to update the changelog just before we make a release. In the long run, there's definitely a good opportunity to have this whole thing automated, so that we only provide the version number once, to a workflow, and it does the whole thing for us: compile the changelog, commit the change, tag that commit, and go through all the other steps of creating and testing a release based on the tag. But I think it needs a good amount of thought to figure out how best to do that, and I'd rather we leave it to discuss after we get at least one release in the books.

So yeah, #90 is my proposal for how we move forward for now - but I'm definitely open to discussion. If you want to take a crack at implementing something a little more sophisticated that will give us a smoother release workflow, go for it - it's just my preference to pause here, finish up our first release, and then come back to that problem.

**Comment by sjlongland** [comment]
Ehh fun, chicken-and-egg problem.  `towncrier` needs the version number, which is derived from the `git` tag, which we want to include the current change-log information but can't because the tag doesn't exist yet.

I think a manual release step seems the most appropriate.  While I'm used to manually updating a version number in the code (NodeJS it's in `package.json`; lots of legacy Python projects have it either in a module property or in `setup.py`), it's not exactly an elegant way of doing this.

I agree it's not a solution that needs immediate fixes, so we can re-visit this later.

**Comment by sjlongland** [comment]
The changes look okay, but I'll just manually run it here and see what it does… make sure it actually generates sensible output.

**Comment by sjlongland** [comment]
```
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ ls -l CHANGELOG.rst 
-rw-r--r-- 1 stuartl stuartl 0 Oct  5 17:32 CHANGELOG.rst
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ python3 -m towncrier build --version 1.0.0
Loading template...
Finding news fragments...
Rendering news fragments...
Writing to newsfile...
Staging newsfile...
I want to remove the following files:
/home/stuartl/projects/setuptools-pyproject-migration/work/changelog.d/88.feature
Is it okay if I remove those files? [Y/n]: n
Done!
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ git status
On branch towncrier/1/dev
Your branch is up to date with 'origin/towncrier/1/dev'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   CHANGELOG.rst

RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ git diff
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ cat CHANGELOG.rst 
1.0.0
=====

Features
--------

- Enable towncrier (#88)
```

So that added a new entry to the previously empty `CHANGELOG.rst`.

---

If I let it delete the file…

```
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ python3 -m towncrier build --version 1.0.0
Loading template...
Finding news fragments...
Rendering news fragments...
Writing to newsfile...
Staging newsfile...
I want to remove the following files:
/home/stuartl/projects/setuptools-pyproject-migration/work/changelog.d/88.feature
Is it okay if I remove those files? [Y/n]: y
Removing news fragments...
Done!
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ git status
On branch towncrier/1/dev
Your branch is up to date with 'origin/towncrier/1/dev'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   CHANGELOG.rst
        deleted:    changelog.d/88.feature
```

…the deletion is staged.  Presumably it then lives _permanently_ in `CHANGELOG.rst` from then on.

So it looks to be doing the right thing when run directly.

---

Running via `tox`:

```
towncrier: commands[0]> towncrier build --yes
Traceback (most recent call last):
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/bin/towncrier", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/towncrier/build.py", line 123, in _main
    return __main(
           ^^^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/towncrier/build.py", line 162, in __main
    project_version = get_version(
                      ^^^^^^^^^^^^
  File "/home/stuartl/projects/setuptools-pyproject-migration/work/.tox/towncrier/lib/python3.11/site-packages/towncrier/_project.py", line 46, in get_version
    raise Exception("No __version__, I don't know how else to look")
Exception: No __version__, I don't know how else to look
towncrier: exit 1 (0.43 seconds) /home/stuartl/projects/setuptools-pyproject-migration/work> towncrier build --yes pid=105672
.pkg: _exit> python /home/stuartl/projects/setuptools-pyproject-migration/env/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  towncrier: FAIL code 1 (19.91=setup[19.48]+cmd[0.43] seconds)
  evaluation failed :( (20.09 seconds)
```

**Comment by sjlongland** [comment]
A ha… so the answer is, we manually provide it via `tox`… otherwise `towncrier` can't figure out the version number:

https://github.com/twisted/towncrier/issues/320

```
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ tox -e towncrier -- --version 1.0.0
.pkg: _optional_hooks> python /home/stuartl/projects/setuptools-pyproject-migration/env/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_editable> python /home/stuartl/projects/setuptools-pyproject-migration/env/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_editable> python /home/stuartl/projects/setuptools-pyproject-migration/env/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
towncrier: install_package> python -I -m pip install --force-reinstall --no-deps /home/stuartl/projects/setuptools-pyproject-migration/work/.tox/.tmp/package/65/setuptools_pyproject_migration-0.0.1.dev4+g16cfb26-0.editable-py3-none-any.whl
towncrier: commands[0]> towncrier build --yes --version 1.0.0
Loading template...
Finding news fragments...
Rendering news fragments...
Writing to newsfile...
Staging newsfile...
Removing the following files:
/home/stuartl/projects/setuptools-pyproject-migration/work/changelog.d/88.feature
Removing news fragments...
Done!
.pkg: _exit> python /home/stuartl/projects/setuptools-pyproject-migration/env/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  towncrier: OK (1.50=setup[1.20]+cmd[0.30] seconds)
  congratulations :) (1.63 seconds)
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ git status
On branch towncrier/1/dev
Your branch is up to date with 'origin/towncrier/1/dev'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   CHANGELOG.rst
        deleted:    changelog.d/88.feature

RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ git diff
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ cat CHANGELOG.rst 
1.0.0
=====

Features
--------

- Enable towncrier (#88)
```

**Comment by diazona** [comment]
Oh when running with tox you have to run it with the `--version` option.

```console
$ tox run -e towncrier -- --version 1.0.0
```

or such. I figured we can make it automatically figure out the version number in the future, but for now it was easiest to just pass it in manually.

**Comment by diazona** [comment]
Haha we got it simultaneously :smile:

**Comment by sjlongland** [comment]
> Haha we got it simultaneously 😄

Classic race condition. :-)
